src/nvim/eval.c

```diff
@@ -11686,14 +11687,61 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
         return;
       }
     }
+    dictitem_T *item;
+    // or new_env
+    item = tv_dict_find(job_opts, S_LEN("env"));
+    ILOG("Just looked into job_opts");
+    if (item) {
+      ILOG("env passed correctly");
+      if (item->di_tv.v_type != VAR_DICT) {
+        EMSG2(_(e_invarg2), "Expected a dictionary for 'env'");
+        return;
+      }
+      // TODO add the following in tv_to_argv  ?
+      //tv_to_argv(&argvars[0], NULL, &executable);
+      size_t count = 0;
+      TV_DICT_ITER(item->di_tv.vval.v_dict, var, {
+          const char *str = tv_get_string(&var->di_tv);
+          count++;
+          if (str) {
+            ILOG("key %s", str);
+            if (env == NULL) {
+              env = xmalloc((size_t)1);
+            } else {
+              env = xrealloc(env, count);
+            }
+            // env[count-1] = xmalloc( (size_t)(STRLEN(di->di_key) + 1 + (STRLEN(str))  ) * sizeof(char_u));
+            // int res = sprintf(env[count-1], "%s=%s", var->di_key, str));
+            // TODO need to escape the var 
+            // use shell_xescape_xquote ?
+            env[count-1] = (char *)concat_str(var->di_key, concat_str((char_u *)"=", (char_u*)(str)));
```

Comment:
Why is any escaping required when constructing the environment?  If anything, the consumer of the environment variable should be ready to deal with whitespace and the like, but as far as I know when setting you just need `key=var` format.